tests/test_pyfdb.py

```diff
@@ -0,0 +1,190 @@
+    "date": "20191110",
+    "time": "0000",
+    "step": "0",
+    "param": "138",
+    "class": "rd",
+    "type": "an",
+    "expver": "xxxx",
+}
+
+
+def test_archival_read():
+
+    filename = os.path.join(os.path.dirname(__file__), "x138-300.grib")
+    pyfdb.archive(open(filename, "rb").read())
+
+    key["levelist"] = "400"
+    filename = os.path.join(os.path.dirname(__file__), "x138-400.grib")
+    pyfdb.archive(open(filename, "rb").read())
+
+    key["expver"] = "xxxy"
+    filename = os.path.join(os.path.dirname(__file__), "y138-400.grib")
+    pyfdb.archive(open(filename, "rb").read())
+    pyfdb.flush()
+
+    # List
+    request = {
+        "class": "rd",
+        "expver": "xxxx",
+        "stream": "oper",
+        "date": "20191110",
+        "time": "0000",
+        "domain": "g",
+        "type": "an",
+        "levtype": "pl",
+        "step": 0,
+        "levelist": [300, "500"],
+        "param": ["138", 155, "t"],
+    }
+    print("direct function, request as dictionary:", request)
+    for el in pyfdb.list(request, True):
+        assert el["path"]
+        assert el["path"].find("rd:xxxx:oper:20191110:0000:g/an:pl.") != -1
+        assert "keys" not in el
+
+    request["levelist"] = ["100", "200", "300", "400", "500", "700", "850", "1000"]
+    request["param"] = "138"
+    print("")
+    print("direct function, updated dictionary:", request)
+    it = iter(pyfdb.list(request, True, True))
+
+    el = next(it)
+    assert el["path"]
+    assert el["path"].find("rd:xxxx:oper:20191110:0000:g/an:pl.") != -1
+    assert el["keys"]
+    keys = el["keys"]
+    assert keys["class"] == "rd"
+    assert keys["levelist"] == "300"
+
+    el = next(it)
+    assert el["path"]
+    assert el["path"].find("rd:xxxx:oper:20191110:0000:g/an:pl.") != -1
+    assert el["keys"]
+    keys = el["keys"]
+    assert keys["class"] == "rd"
+    assert keys["levelist"] == "400"
+
+    try:
+        el = next(it)
+        assert False, "returned unexpected field"
+    except StopIteration:
+        assert True, "field listing completed"
+
+    # as an alternative, create a FDB instance and start queries from there
+    request["levelist"] = ["400", "500", "700", "850", "1000"]
+    print("")
+    print("fdb object, request as dictionary:", request)
+    for el in fdb.list(request, True, True):
+        assert el["path"]
+        assert el["path"].find("rd:xxxx:oper:20191110:0000:g/an:pl.") != -1
+        assert el["keys"]
+        keys = el["keys"]
+        assert keys["class"] == "rd"
+        assert keys["levelist"] == "400"
+
+    # Retrieve
+    request = {
+        "domain": "g",
+        "stream": "oper",
+        "levtype": "pl",
+        "step": "0",
+        "expver": "xxxx",
+        "date": "20191110",
+        "class": "rd",
+        "levelist": "300",
+        "param": "138",
+        "time": "0000",
+        "type": "an",
+    }
+
+    filename = os.path.join(os.path.dirname(__file__), "x138-300bis.grib")
+    print("")
```

Comment:
These tests seem to be very ... noisy/chatty

Reply:
This still is the initial test.